igor-web/src/main/groovy/com/netflix/spinnaker/igor/build/BuildController.groovy

```diff
@@ -51,205 +55,238 @@ import static org.springframework.http.HttpStatus.NOT_FOUND
+    def job = ((String) request.getAttribute(
+      HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE)).split('/').drop(4).join('/')
+    def buildService = getBuildService(master)
+    return buildService.getBuilds(job)
+  }
+
+  @RequestMapping(value = "/masters/{name}/jobs/{jobName}/stop/{queuedBuild}/{buildNumber}", method = RequestMethod.PUT)
+  @PreAuthorize("hasPermission(#master, 'BUILD_SERVICE', 'WRITE')")
+  String stop(
+    @PathVariable("name") String master,
+    @PathVariable String jobName,
+    @PathVariable String queuedBuild,
+    @PathVariable Integer buildNumber) {
+
+    def buildService = getBuildService(master)
+    if (buildService instanceof JenkinsService) {
+      // Jobs that haven't been started yet won't have a buildNumber
+      // (They're still in the queue). We use 0 to denote that case
+      if (buildNumber != 0 &&
+        buildService.metaClass.respondsTo(buildService, 'stopRunningBuild')) {
+        buildService.stopRunningBuild(jobName, buildNumber)
+      } else {
+        // The jenkins api for removing a job from the queue (http://<Jenkins_URL>/queue/cancelItem?id=<queuedBuild>)
+        // always returns a 404. This try catch block insures that the exception is eaten instead
+        // of being handled by the handleOtherException handler and returning a 500 to orca
+        try {
+          if (buildService.metaClass.respondsTo(buildService, 'stopQueuedBuild')) {
+            buildService.stopQueuedBuild(queuedBuild)
+          }
+        } catch (RetrofitError e) {
+          if (e.response?.status != NOT_FOUND.value()) {
+            throw e
+          }
+        }
+      }
     }
 
-    @RequestMapping(value = '/builds/all/{master:.+}/**')
-    @PreAuthorize("hasPermission(#master, 'BUILD_SERVICE', 'READ')")
-    List<Object> getBuilds(@PathVariable String master, HttpServletRequest request) {
-        def job = ((String) request.getAttribute(
-            HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE)).split('/').drop(4).join('/')
-        def buildService = getBuildService(master)
-        return buildService.getBuilds(job)
+    "true"
+  }
+
+  @RequestMapping(value = '/masters/{name}/jobs/**', method = RequestMethod.PUT)
+  @PreAuthorize("hasPermission(#master, 'BUILD_SERVICE', 'WRITE')")
+  String build(
+    @PathVariable("name") String master,
+    @RequestParam Map<String, String> requestParams, HttpServletRequest request) {
+    def job = ((String) request.getAttribute(
+      HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE)).split('/').drop(4).join('/')
+
+    String pendingKey = computePendingBuildKey(master, job, requestParams)
+    String buildNumber = null
+
+    PendingOperationsCache.OperationState pendingStatus = pendingOperationsCache.getAndSetOperationStatus(pendingKey, PendingOperationsCache.OperationStatus.PENDING, "")
+    if (pendingStatus.status == PendingOperationsCache.OperationStatus.PENDING) {
+      throw new BuildOperationPendingException(master, job)
```

Comment:
Throwing this exception like this initially struck me as odd - curious for the reasoning?  Instead of throwing an exception is it possible to just return an accepted response with a message?

Reply:
Yeah, I actually hate it - but I couldn't figure out how to return `202-ACCEPTED` if you can point me in the right direction i would love to change it as it will reduce the noise in logs

Reply:
nvm, @srekapalli put me on the path of righteousness :)

Reply:
fixed